### ISSUE TITLE
bug in ref_batch_normalization.cpp ?

### ISSUE THREAD

**Issue by zeno40** [intel/mkl-dnn#517]
Hi,

In ref_batch_normalization.cpp at
line 139:   v_variance /= W * H * N * D;
must be:    v_variance /= W * H * N * D - 1;
according to the original paper and implementation.

thanks

**Comment by rsdubtso** [comment]
I'm looking at [this paper](https://arxiv.org/pdf/1502.03167.pdf) and cannot find the `-1` term anywhere. Can you please point me to the paper and implementation you mentioned?

**Comment by zeno40** [comment]
in tiny-dnn for example the function to calculate the moments of BN at line 57-60.

https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/util/math_functions.h

You don't gonna find a  N*D*H*W-1 in the ref paper. A bit stupid of me to make such statement.
But calculating the variance in this way was explained on many blog many years ago.  There was a good reason for it if I remember correctly.  But I don't remember the reason for it clearly.

**Comment by rsdubtso** [comment]
I can clearly see `-1` on [line 60](https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/util/math_functions.h#L60). Maybe the difference is between using [sample variance](https://en.wikipedia.org/wiki/Variance#Sample_variance) (or [moments](https://en.wikipedia.org/wiki/Moment_(mathematics)#Sample_moments)) vs [population moments](https://en.wikipedia.org/wiki/Variance#Population_variance)?

* FWIW, TensorFlow seems to be [using population variance in `nn.moments`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_impl.py#L1194) since it uses the same `math_ops.reduce_mean` for both mean and variance statistics.
* Caffe seems to be [using sample statistics](https://github.com/BVLC/caffe/blob/master/src/caffe/layers/batch_norm_layer.cpp#L142)
* If you have pointers on how other frameworks implement BN, I'd be happy to take a look.

The batch normalization paper mentions sample statistics when discussing inference (https://arxiv.org/pdf/1502.03167.pdf, page 4, second column, paragraph after the formula):

> We use the unbiased variance estimate Var[x] = \frac{m}{m-1} ·E_B[σ_B^2], where the expectation is over training mini-batches of size m and σ_B^2 are their sample variances.

But the original formulae are definitely using population statistics.

**Comment by zeno40** [comment]
in darknet: https://github.com/pjreddie/darknet/blob/master/src/blas.c
line 110-
void variance_cpu(float *x, float *mean, int batch, int filters, int spatial, float *variance)
float scale = 1./(batch * spatial - 1);

**Comment by zeno40** [comment]
I'm just asking because I extensively use mkl-dnn and I like it because it's very fast! But I have to write a custom fused batchnorm + swish layer for reducing the memory load during training of a mobilenetv3.
So I took ref_batch_normalization.cpp as a primary example.

If TensorFlow uses population variance like mkl-dnn I will just do the same. Hard to imagine
the folks how wrote TensorFlow or mkl-dnn made an error.

**Comment by rsdubtso** [comment]
I am also trying to figure out what's right. We (mkl-dnn) might very well have missed something!

FWIW, if you are using MKL-DNN, it already supports swish (#501) and should probably support merged BN+Swish too.

**Comment by zeno40** [comment]
In torch they use also the unbiased varianc. see line 48

https://github.com/torch/nn/blob/master/lib/THNN/generic/BatchNormalization.c#L22

**Comment by rsdubtso** [comment]
> In torch they use also the unbiased variance, see line 48
> 
> https://github.com/torch/nn/blob/master/lib/THNN/generic/BatchNormalization.c#L22

These are running average and variance. They are expected to use sample forms. I don't think mkl-dnn computes those for you (#371).

**Comment by zeno40** [comment]
This is indeed the calculation of the running mean/variance, but these are calculated from the actual mean of the batch and in this case the unbiased variance of the batch. So, there doing the same as in the darknet implementation of BN. I know mkl-dnn doesn't calculate the running stats but don't we have to
base them on the unbiased variance instead?

**Comment by zeno40** [comment]
I think I found the culprit for the confusion. The implementation of ref_batch_normalization.cpp is correct because during training you must indeed divide by NDHW to calculate the variance you need but at the same time during training we also need the unbiased variance for calculating the moving averages but currently we only get the variance out of the primitive and not the unbiased variance we actually need for our calculation of the running variance.

**Comment by rsdubtso** [comment]
Thanks. I'll update the documentation and close this.
